### ISSUE TITLE
Subspecs should result in separate frameworks

### ISSUE THREAD

**Issue by fluidsonic** [CocoaPods/CocoaPods#3727]
TLDR: CocoaPods should create separate frameworks for each Subspec.

------------------------------

We have a larger Swift framework we'd like to split into a lot Subspecs.

The problem is that with `use_frameworks!` the source files of all Subspecs are thrown into a single framework, which causes several issues:
- it destroys the Pod's intentional structure
- it causes conflicts since all Swift files are located in the same module
- it causes build errors when two different Subspecs have a Swift file with the same name

For example using the following Podspec configuration:
- MyPod
  - A
  - B
    - C
    - D

should not result in
```Swift
import MyPod
```
but
```Swift
import MyPod_A
import MyPod_B_C
import MyPod_B_D
```

Actually we'd like to have a true module hierarchy but the following is not (yet?) possible in Swift:
```Swift
import MyPod.A
import MyPod.B.C
import MyPod.B.D
```

**Comment by segiddins** [comment]
> Actually we'd like to have a true module hierarchy but the following is not (yet?) possible in Swift

If you want that, you will need to manually create & specify a custom module map file in your Podspec.

**Comment by fluidsonic** [comment]
I experimented a lot with module maps today. The problem is that clang supports submodules while Swift does not. I.e. the compiler doesn't even look for `.swiftmodule` files of submodules. The `import Pod.Subspec` would not fail but that's about it. The submodule's symbols are never loaded.
So it looks like true submodules are impossible at the moment. I filed bug 21528060 for that.

Creating separate frameworks using a naming scheme like `PodName_SubspecName` doesn't rely on clang submodules and thus should be possible.

**Comment by segiddins** [comment]
> Creating separate frameworks using a naming scheme like PodName_SubspecName doesn't rely on clang submodules and thus should be possible.

That would horribly break pretty much every existing pod. 

And using custom module maps to create explicit submodules works, we use it at Realm.
If you actually want things to be in separate frameworks, they need to be in separate pods.

**Comment by fluidsonic** [comment]
Most existing Pods don't have Subspecs and thus won't break.
I'm aware that there are Pods with Subspecs which are already being used by projects using `use_frameworks!` so in any case this behavior would have to be opt-in, sure.

If you refer to [realm-cocoa's modulemap](https://github.com/realm/realm-cocoa/blob/master/Realm/module.modulemap) then this is not a true submodule.

Realm is one single big module named `Realm` where **some of it's symbols are just hidden** behind the *Realm.Private* submodule (for the consumer of the framework).
More importantly Realm is written in Objective-C. I'm writing about Swift modules and there are no header files in Swift you could hide using this trick.

Realm is using [inferred submodules](http://clang.llvm.org/docs/Modules.html#submodule-declaration) just to control symbol visiblity.
What I was writing about is independent *nested modules* where the code for each module is isolated.
Right now the Swift code of each Subspec can access the symbols of all other Subspecs since all of them are compiled in the same module. That breaks the separation of concerns principle and is a source for errors where Subspecs can accidentally use each other's code without intention. This also causes collision of symbols and file names.

Without such isolation I'd say Subspecs don't make much sense. It's just a control meachnism where the consumer of a Pod can add/remove some of the Pod's source files by adding/removing Subspecs.

Sure we could create separate Pods for each but why do we have Subspecs then in the first place?

In the current situation this will lead to one of two scenarios:
- Big Pods where the code is all compiled together with zero isolation.
Maintaining a good structure there is awful, esp. when your Pod has a lot small but isolated functionality which may or may not depend on each other.
- Many small Pods which built upon each other although it barely makes sense to use most of them on their own. That way the Pods' creator can achieve good code quality by isolating functionality and the consumer of the Pod can selectively `import` functionality instead of `import`ing all functionality at once. But that would lead to spammy `MyPod_MySubspec_SomeInternalSubspec` Pods.

We only learned how much we could isolate & improve our code when we began splitting a big app into separate Pods and then just plug these together. All parts are isolated yet belong together somehow as they are part of the same app. They're still separate Pods because there's no other way to achieve that.

----------------

I don't see how creating separate frameworks per Subspec could cause anything but improved code quality for both, the Pod creator and its consumers.

**Comment by segiddins** [comment]
We have sub specs because of legacy, I'd never personally advocate using them.

**Comment by fluidsonic** [comment]
What's the long-term plan then? Just keep them alive until they eventually die?
They could become a great way for everyone to improve the structure of Pods & apps.
